integration/client-data-test.ts

```diff
@@ -0,0 +1,357 @@
+      export function loader() {
+        return json({ message: 'Parent Server Loader'});
+      }
+      ${parentAdditions}
+      export default function Component() {
+        let data = useLoaderData();
+        return (
+          <>
+            <h1 id="parent-heading">Parent Component</h1>
+            <p id="parent-data">{data.message}</p>
+            <Outlet/>
+          </>
+        );
+      }
+    `,
+    "app/routes/parent.child.tsx": js`
+      import { json } from '@remix-run/node'
+      import { Outlet, useLoaderData } from '@remix-run/react'
+      export function loader() {
+        return json({ message: 'Child Server Loader'});
+      }
+      ${childAdditions}
+      export default function Component() {
+        let data = useLoaderData();
+        return (
+          <>
+            <h2 id="child-heading">Child Component</h2>
+            <p id="child-data">{data.message}</p>
+            <Outlet/>
+          </>
+        );
+      }
+    `,
+  };
+}
+
+test.describe("Client Data", () => {
+  let appFixture: AppFixture;
+
+  test.afterAll(() => {
+    appFixture.close();
+  });
+
+  test.describe("Initial Hydration", () => {
+    test("no client loaders or fallbacks", async ({ page }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({ parentAdditions: "", childAdditions: "" }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Full SSR - normal Remix behavior due to lack of clientLoader
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Component");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Component");
+      expect(html).toMatch("Child Server Loader");
+    });
+
+    test("parent.clientLoader/child.clientLoader", async ({ page }) => {
+      appFixture = await createAppFixture(
+        await createFixture({
+          files: getFiles({
+            parentAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Parent Client Loader" };
+            }
+          `,
+            childAdditions: js`
+            export async function clientLoader() {
+              await new Promise(r => setTimeout(r, 100));
+              return { message: "Child Client Loader" };
+            }
+          `,
+          }),
+        })
+      );
+      let app = new PlaywrightFixture(appFixture, page);
+
+      // Full SSR - normal Remix behavior due to lack of Fallback components
+      await app.goto("/parent/child");
+      let html = await app.getHtml("main");
+      expect(html).toMatch("Parent Component");
+      expect(html).toMatch("Parent Server Loader");
+      expect(html).toMatch("Child Component");
+      expect(html).toMatch("Child Server Loader");
```

Comment:
In this case, I understand the flow will be
1. SSR with server loaders
2. Hydration with server data
And client loader will be ignored until the user navigates back to this route client-side only?

Reply:
yeah - `clientLoader`'s are _always_ called on SPA navigations.  They are skipped on initial hydration by default, and called on hydration when a `Fallback` exists.